### ISSUE TITLE
GitHub Enterprise Support 

### ISSUE THREAD

**Issue by czechboy0** [czechboy0/Buildasaur#99]
Let's add support for self-hosted GitHub. Unfortunately, I don't have an Enterprise instance running anywhere, so it'd be very hard to test with. If someone is interested in helping me develop this feature for Buildasaur, I will need a couple of things:
- [ ] A dummy git repo hosted on your Enterprise instance, can be just a repo of a template Xcode project
- [ ] A GitHub token that can access that one repo
- [ ] SSH keys that can access that one repo (or preferably, it could be open completely, so no git authentication is required)

**Comment by czechboy0** [comment]
Work in progress: https://github.com/czechboy0/Buildasaur/compare/gladiolus...hd/github_enterprise?expand=1

And now I'm basically blocked by the things above. Without them, I can't really continue.

**Comment by rcaileff** [comment]
I've taken a shot at implementing this:
https://github.com/czechboy0/Buildasaur/pull/248

**Comment by rcaileff** [comment]
Different implementation based on comments to the first one.
https://github.com/czechboy0/Buildasaur/pull/251

**Comment by fotiDim** [comment]
Voting for SSH keys instead of personal tokes.

**Comment by rcaileff** [comment]
Does the implementation in https://github.com/czechboy0/Buildasaur/pull/248 work for you, or did you have a third alternative in mind?

**Comment by fotiDim** [comment]
I didn't try it yet because the PR was closed so I guessed it wasn't ready. Should I go ahead?

**Comment by rcaileff** [comment]
It's the version I'm using for my team.  If it is a better form of
authentication for you, you should use it.

-Rachel

**Comment by fotiDim** [comment]
@rcaileff  just tried it and I cannot past this
![screen shot 2016-03-18 at 15 59 07](https://cloud.githubusercontent.com/assets/2326415/13883296/168df52c-ed29-11e5-8a79-2e39dd95a92c.png)

I have done the workaround but still nothing.

PS: I am using your "enterprise_github" branch.

**Comment by rcaileff** [comment]
Do you have a lounge.xcworkspace/xcshareddata/lounge.xcscmblueprint
blueprint file with the string "http" in it?
Did you clone your Lounge repo using github's "SSH" link or the "HTTPS"
link?

**Comment by fotiDim** [comment]
I cloned using the SSH url. I have setup deploy Keys for authentication. My remote is in the "git@github.my.custom.url:organization/ios-app.git" form and within the .xcscmblueprint file (in the path you mentioned) I have been trying all combinations. This is it a bit altered:

`{
    "DVTSourceControlWorkspaceBlueprintPrimaryRemoteRepositoryKey": "AA2A0A40BEDCE6DCA89ECA",
    "DVTSourceControlWorkspaceBlueprintWorkingCopyRepositoryLocationsKey": {},
    "DVTSourceControlWorkspaceBlueprintWorkingCopyStatesKey": {
        "AA2A0A40BEDCE6DCA89ECA": 0
    },
    "DVTSourceControlWorkspaceBlueprintIdentifierKey": "987E0-4183-8E69-DD84FE4BEAAD",
    "DVTSourceControlWorkspaceBlueprintWorkingCopyPathsKey": {
        "AA2A0A40BEAA46034CE6DCA89ECA": "Folder/"
    },
    "DVTSourceControlWorkspaceBlueprintNameKey": "app",
    "DVTSourceControlWorkspaceBlueprintVersion": 204,
    "DVTSourceControlWorkspaceBlueprintRelativePathToProjectKey": "app.xcworkspace",
    "DVTSourceControlWorkspaceBlueprintRemoteRepositoriesKey": [
        {
            "DVTSourceControlWorkspaceBlueprintRemoteRepositoryURLKey": "git@github.my.custom.url:organization/ios-app.git",
            "DVTSourceControlWorkspaceBlueprintRemoteRepositorySystemKey": "com.apple.dt.Xcode.sourcecontrol.Git",
            "DVTSourceControlWorkspaceBlueprintRemoteRepositoryIdentifierKey": "AA2A0A40BEDCE6DCA89ECA"
        }
    ]
}`

**Comment by rcaileff** [comment]
I don't see a problem with the blueprint file.  What about the link used to
make the test machine's clone of the repo?  Was it the SSH link or the
HTTPS link?

**Comment by fotiDim** [comment]
I used the SSH link for all operations. Notice that it starts with git@ and not with ssh://

**Comment by rcaileff** [comment]
The blueprint file contents and the repo cloning method don't necessarily
match.  That was a mistake I made when setting up buildasaur the first
time.  :)

**Comment by rcaileff** [comment]
Is your EnterpriseGitHubHostname pod key value set to the same
github.my.custom.url that appears in your blueprint file?

**Comment by fotiDim** [comment]
@rcaileff good question, I'll have to check. I also noticed some letfovers from an old remote URL inside the .xccheckout file. I'll try to test on a fresh project to make sure. I am under intranet so I will have to wait until Monday for this.

**Comment by fotiDim** [comment]
@rcaileff I have been trying to figure it out since morning. Seems that the problem is at WorkspaceMetadata.swift at line 85 and 88:
`switch scheme {
        ...
        case GitService.EnterpriseGitHub.hostname():
            return (CheckoutType.SSH, .EnterpriseGitHub)
        ...
`

my scheme variable is 
`"github._subdomain_._domain_.com"`
while GitService.EnterpriseGitHub.hostname() returns `"git@github._subdomain_._domain_.com:organisation/ios-app.git"`
Is it a misconfiguration from my side?

**Comment by rcaileff** [comment]
GitService.EnterpriseGitHub.hostname() should be returning the value from
your EnterpriseGitHubHostname pod key, which is why I asked about it.

**Comment by fotiDim** [comment]
Finally I got past this step. I just had to change the pod key value to "github._subdomain_._domain_.com" :/

So now I am on this screen and a bit confused 
<img width="441" alt="screen shot 2016-03-22 at 17 39 11" src="https://cloud.githubusercontent.com/assets/2326415/13959597/07b9835c-f055-11e5-9988-d127dd279215.png">

I have set up ssh authentication already. This means that I can pull/push though git without providing a password.

I have pointed buildasaur to my private SSH key. Do I need to perform oauth authentication as well? If yes then when I click on the login button and enter my credentials I am redirect to a 404 page. The URL that it points to is this: `https://github.subdomain.domain.com/login/oauth/authorize?client_id=a&redirect_uri=buildasaur%3A%2F%2Foauth-callback%2Fgithub&response_type=code&scope=repo&state=30N9BUAJZkrIp7yoO06t`

Any ideas?

Keep in mind that I haven't set a client ID and secret for oauth to pod keys.

**Comment by rcaileff** [comment]
The version in #248 (or in my works_for_me or enterprise_github branches of
my repo) requires using OAuth.  Using it requires registering an OAuth
application with an Authorization callback URL of
"buildasaur://oauth-callback/github".
The version in #251 does not require using OAuth.  Using it requires
generating a Personal Access Token.

**Comment by fotiDim** [comment]
OK got past this step as well. Unfortunately I stumble upon my old problem (same as with personal token). We are using SAML for authentication and this takes Buildasaur to the HTML page to enter the SAML credentials before being let into GitHub.

Buildosaur app is granted access on GitHub
![screen shot 2016-03-23 at 13 51 44](https://cloud.githubusercontent.com/assets/2326415/13985705/7fe5e966-f0fe-11e5-9be2-70970c26082d.png)

Strange thing is that Buildosaur says it is logged in.
![screen shot 2016-03-23 at 13 39 00](https://cloud.githubusercontent.com/assets/2326415/13985965/09c96be8-f100-11e5-95d7-3c3ff759bf58.png)

PS: I am trying to find a way to bypass this step but so far our IT guys say it is not possible.

**Comment by rcaileff** [comment]
Which error message are you currently getting?

**Comment by fotiDim** [comment]
On the console I am getting only this: "[ERROR]: No X-RateLimit info provided by GitHub in headers: [Pragma: no-cache, Content-Type: text/html;charset=ISO-8859-1, Server: Apache-Coyote/1.1, Date: Thu, 24 Mar 2016 14:47:45 GMT, Content-Length: 1838, Cache-Control: no-cache, no-store, must-revalidate, max-age=0, Expires: 0], we're unable to detect the remaining number of allowed requests. GitHub might fail to return data any time now :("

but as you see on the screenshot I get "Failed to access server, error:" and then the HTML of the SAML login page.

**Comment by rcaileff** [comment]
Here's my best guess at what's happening:
In GitHubServer.swift the private _sendRequest function is parsing header data that your server doesn't provide.  It looks like you'd need to change what headers Buildasaur expects or change what headers the server is sending to get past your current problem.
If that's not it, I have no idea what's wrong.  :(

**Comment by czechboy0** [comment]
Yeah for GitHub Buildasaur expects rate limit information in the headers. You could make that optional for EnterpriseGitHub which should hopefully fix the first issue.
The second one seems to be that the API Buildasaur is calling when verifying credentials is returning HTML instead of JSON for some reason. Hopefully that should be easy to debug with an attached debugger.

**Comment by rcaileff** [comment]
Why does GitHub Buildasaur expect that information?  Is it safe to remove
it for Enterprise GitHub?  If so, why keep it for regular GitHub?

**Comment by czechboy0** [comment]
So that when you're using too short of a sync interval and run out of the allowed 5k requests per hour, Buildasaur will tell you what happened. Otherwise it'd just start failing without telling you it was gonna happen. 

If there are no limits for Enterprise GitHub, feel free to ignore it there. But regular GitHub needs it.
